python/ray/data/datasource/tfrecords_datasource.py

```diff
@@ -32,11 +35,82 @@ class TFRecordDatasource(FileBasedDatasource):
     def _read_stream(
         self, f: "pyarrow.NativeFile", path: str, **reader_args
     ) -> Iterator[Block]:
+        tf_schema: Optional["schema_pb2.Schema"] = reader_args.get("tf_schema", None)
+        batch_size = reader_args.get("batch_size", 2048)
+
+        if tf_schema:
+            try:
+                self._fast_read_stream(path, tf_schema, batch_size)
+                return
+            except ModuleNotFoundError as e:
+                print(f"Failed to import tfx_bsl; falling back to slow read: {e}")
+        self._slow_read_stream(f, path, tf_schema, batch_size)
+
+    def _fast_read_stream(
+        self, path: str, tf_schema: "schema_pb2.Schema", batch_size=2048
+    ):
+        import pyarrow as pa
+
+        try:
+            from tfx_bsl.cc.tfx_bsl_extension.coders import ExamplesToRecordBatchDecoder
+        except ModuleNotFoundError:
+            raise ModuleNotFoundError(
+                "To use CustomTFRecordDatasource, please install "
+                "tfx-bsl package with `pip install tfx-bsl`."
+            )
+
+        decoder = ExamplesToRecordBatchDecoder(tf_schema)
+        for record in tf.data.TFRecordDataset(path, compression_type="GZIP").batch(
```

Comment:
also, passing the path here will work for local filesystems, it won't work for remote file systems such as `gcs`, since the path won't contain the full path